CORE-14880: Migrate "OIDC" Admin API v1 Security Endpoints to v2 ConnectRPC API

nguyen-andrew opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR introduces the security service for Redpanda Admin API v2. Starting with the OIDC-related RPCs, this v2 service initially aims to achieve functional parity with the v1 API by providing the following:
- OIDC Identity Resolution - Resolves OIDC token identity and returns principal mapping with token expiration
- Cluster-wide OIDC Key Refresh - Refreshes OIDC keys across all brokers in the cluster
- Cluster-wide OIDC Session Revocation - Refreshes OIDC keys and revokes all active OIDC sessions cluster-wide

The implementation has the following characteristics:
- For OIDC key refresh and ODIC session revocation, any broker can receive requests and will broadcast the operation across all nodes in the cluster.
- For broadcasted/proxied operations, any error produced by a proxied request is returned to the client
- Session revocation combines key refresh with active session termination

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.
-->

### Features
* Added OIDC security service to Admin API v2 with identity resolution, cluster-wide key refresh, and session revocation capabilities. Administrators can now manage OIDC authentication security through the new v2 endpoints with proper cluster coordination and error handling.

<!--
### Improvements

* Short description of how this PR improves existing behavior.

-->

nguyen-andrew committed:
proto: Adding security.proto for Admin API v2

Creating a new proto file for security-related messages in Admin API v2.
This new proto initially includes definitions for OIDC-related services
and messages.

nguyen-andrew committed:
admin: Implement OIDC security service stubs

This commit adds the skeleton implementation of the OIDC security service
for the Redpanda Admin API v2. The service includes unimplemented stubs for
the following methods:
- resolve_oidc_identity
- refresh_oidc_keys
- revoke_oidc_sessions

nguyen-andrew committed:
admin: OIDC identity resolution in Admin API v2

Add OIDC identity resolution to Admin API v2, updates application
wiring, and extends oauth tests to validate the new functionality.

nguyen-andrew committed:
admin: add refresh_oidc_keys RPC (cluster-wide)

Implement refresh_oidc_keys RPC to refresh OIDC keys by calling
any broker. The receiving broker executes the refresh on all local
shards and proxies/broadcasts the request to all other nodes.
Brokers receiving proxied calls only execute locally and do not
rebroadcast to other nodes. The originating node aggregates failures
across all nodes/shards and returns an aggregated error if any
occurred, otherwise success.

Add tests ensuring a single-node call refreshes OIDC keys across
the cluster.

nguyen-andrew committed:
admin: implement OIDC session revocation in v2 API

Implement RevokeOidcSessions RPC to refresh OIDC keys and revoke all
active OIDC sessions across the cluster. The receiving broker
refreshes OIDC keys on all local shards, revokes OIDC SASL
sessions with kafka::server, and broadcasts the request to all
other nodes in the cluster, with each broker performing local
key refresh and session revocation. The originating node aggregates
errors across all brokers and returns an aggregated error if any
node fails, otherwise success.

Add comprehensive test coverage validating session revocation with
various authentication scenarios. This enables administrators to
immediately invalidate all OIDC sessions when security policies
change or credentials are compromised.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0abd7307-fb00-4b26-b14b-c27a3db3a0e7-->
<!--TEST RESULTS-->
<!--0260aed2-883d-40ee-b966-cd15ba322223-->
<!--TEST RESULTS-->
<!--9cceea2f-2831-4281-b846-90dc660c3a1d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77822">build#77822</a></summary>


| test_class                 | test_method               | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                       |
|:---------------------------|:--------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionReassignmentsTest | test_reassignments_cancel | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77822#019b16a8-20c8-4f42-a2cc-13b1ae5a0406 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1167, p0=0.7110, reject_threshold=0.0100. adj_baseline=0.3109, p1=0.1331, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77928">build#77928</a></summary>


| test_class                 | test_method                                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:---------------------------|:-------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CrashLoopChecksTest        | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77928#019b25c2-c795-4100-8e31-f6edd8f17288 | FAIL          | 0/51     | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.9539, p0=0.0944, reject_threshold=0.0100. adj_baseline=0.9999, p1=1.0000, trust_threshold=0.5000)                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| CrashLoopChecksTest        | test_cloud_storage_client_misconfiguration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77928#019b25c4-18b8-4b9f-a403-6f671ba7ace9 | FLAKY         | 2/21     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.9375, p0=0.8741, reject_threshold=0.0100. adj_baseline=0.9998, p1=0.0000, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CrashLoopChecksTest&test_method=test_cloud_storage_client_misconfiguration |
| DataMigrationsApiTest      | test_higher_level_migration_api            | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77928#019b25c4-18bc-49a9-91cc-ee03a81f05e9 | FLAKY         | 20/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0159, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0471, p1=0.3812, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_higher_level_migration_api          |
| PartitionReassignmentsTest | test_reassignments_cancel                  | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77928#019b25c2-c79c-4ff1-9e2c-076d45e994a5 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0817, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.2256, p1=0.0775, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionReassignmentsTest&test_method=test_reassignments_cancel           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77977">build#77977</a></summary>


| test_class                       | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77977#019b282f-cfb4-4c82-b053-60c28e559bca | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0580, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1642, p1=0.1664, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

nguyen-andrew committed:
dt/pb: Generate Ducktape protobuf clients

Generate Ducktape protobuf clients to test the new Admin API v2 OIDC
security endpoints.

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -0,0 +1,262 @@
+
+    if (!failed.empty()) {
+        serde::pb::rpc::error_info merged_error_info;
+        merged_error_info.reason = "Failed to refresh OIDC keys on some nodes";
```
rpc info reasons should be structured enums and need to follow the pattern: https://github.com/redpanda-data/redpanda/blob/1171d9ee596e08ca45d424d926be84448e97f1c7/src/v/serde/protobuf/rpc.h#L167-L172

nguyen-andrew replied:
Removed

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -0,0 +1,262 @@
+                    }
+                }
+            } catch (...) {
+                vunreachable("Unexpected exception type");
```
we should not crash the server if there is some other error, just rethrow it and it will be logged and converted to internal error

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -0,0 +1,262 @@
+                vunreachable("Unexpected exception type");
+            }
+        }
+        throw serde::pb::rpc::internal_exception(std::move(merged_error_info));
```
Why blanket convert to internal_exception?

nguyen-andrew replied:
Removed

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -0,0 +1,262 @@
+        futures.push_back(client.refresh_oidc_keys(ctx, {}));
+    }
+
+    auto results = co_await ss::when_all(futures.begin(), futures.end());
```
TBH this code is not performance sensitive, I think it would be fine to not do them in parallel and this code would be simpler

nguyen-andrew replied:
That's fair, I've refactored to simplify

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -0,0 +1,262 @@
+          "Failed to refresh OIDC keys on some shards: {}",
+          ex_msg);
+        serde::pb::rpc::error_info info;
+        info.reason = "Failed to refresh OIDC keys";
```
This should be the error message, the reason needs to be structured, and in this case I don't think is useful.

nguyen-andrew replied:
Gotcha, removed

rockwotj commented on proto/redpanda/core/admin/v2/security.proto:
```diff
@@ -0,0 +1,75 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package redpanda.core.admin.v2;
+
+import "google/protobuf/timestamp.proto";
+import "proto/redpanda/core/pbgen/options.proto";
+import "proto/redpanda/core/pbgen/rpc.proto";
+
+option (pbgen.cpp_namespace) = "proto::admin";
+
+service SecurityService {
+    rpc ResolveOidcIdentity(ResolveOidcIdentityRequest)
+        returns (ResolveOidcIdentityResponse) {
+        option (pbgen.rpc) = {
+            authz: USER
+        };
+    }
+
+    rpc RefreshOidcKeys(RefreshOidcKeysRequest)
+        returns (RefreshOidcKeysResponse) {
+        option (pbgen.rpc) = {
+            authz: SUPERUSER
+        };
+    }
+
+    rpc RevokeOidcSessions(RevokeOidcSessionsRequest)
+        returns (RevokeOidcSessionsResponse) {
+        option (pbgen.rpc) = {
+            authz: SUPERUSER
+        };
+    }
+}
```
Public APIs must have documentation. For service, methods and also all proto types. Please see the admin v2 wiki guide on how to write comments from docs team guidance.

nguyen-andrew commented on tests/rptest/tests/redpanda_oauth_test.py:
Found `regenerate_ducktape_protos.sh`, will update test to use the generated clients.

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -11,13 +11,54 @@
+    if (auth_result == nullptr) {
+        vlog(securitylog.warn, "No request_auth_result found in context");
+        throw serde::pb::rpc::internal_exception(
+          "No authentication result found");
```
I don't think a 5xx error is appropriate here.. maybe 412 (precondition failed)

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -11,13 +11,54 @@
+      auth_result->get_sasl_mechanism()
+      != security::oidc::sasl_authenticator::name) {
+        throw serde::pb::rpc::unauthenticated_exception(
+          "SASL mechanism is not OIDC");
```
I also think this should be a 412 (pre-condition failed).  I worry that 'unauthenticated' indicates that the OIDC token was bad (which in this case it isn't).

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -11,13 +11,54 @@
+
+    security::oidc::authenticator auth{_controller->get_oidc_service().local()};
+    auto res = auth.authenticate(
+      auth_result->get_password().substr(authz_bearer_prefix.length()));
```
nit: This seems redundant to re-authenticate the token (which shoudl have already been authenticated) to just get the expiration timeout.  I wonder if the you could either expand the auth_result to also contain that information or maybe expose something in `security` that can do a simple parse of the provided token.

nguyen-andrew replied:
Definitely agree, maybe we can expand auth_result to include an `auth_payload` or something and use that instead instead of having to decode the token again - I'd like to explore that in a follow-on PR if that's okay.

michael-redpanda commented on tests/rptest/tests/redpanda_oauth_test.py:
```diff
@@ -330,6 +331,67 @@ def request_whoami(with_auth: bool):
+        # At this point, admin API does not require auth and service_user_id is not a superuser
+
+        response = request_resolve_oidc_identity(with_auth=False)
+        assert response.status_code == requests.codes.unauthorized
```
chore: python asserts (which raise `AssertionError`) do not really provide enough context.  You should add a message, e.g.

```python
assert response.status_code == requests.codes.unauthorized, f'Unexpected status code: {response.status_code} != {requests.codes.unauthorized}`
```

michael-redpanda replied:
do this for all of your asserts

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/4e7fbc2fe7ad37748e05c8aa824ac24548330de6..46c6bfca472001583d86cb8a476e92501877cf67) to address PR comments:
- Added documentation to the security.proto
- Updated `resolve_oidc_identity` to use `failed_precondition_exception` as suggested.
- Simplified implementations of `refresh_oidc_keys` & `revoke_oidc_sessions`.
- Updated the Admin V2 tests added in `redpanda_oauth_test.py` to use the generated client.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77928-->
<!--2e676392-2b77-460f-bf79-c119ebc33fe4-->
<!--4b64e66e-ae56-4448-a1d4-06239c25dced-->
<!--title-->
## Retry command for [Build#77928](https://buildkite.com/redpanda/redpanda/builds/77928)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/crash_loop_checks_test.py::CrashLoopChecksTest.test_cloud_storage_client_misconfiguration
```
<!--/body-->

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/46c6bfca472001583d86cb8a476e92501877cf67..2577d2f52fddc33954689caa21febe90bacba957) to rebase on latest dev.

nguyen-andrew commented on proto/redpanda/core/admin/v2/security.proto:
```diff
@@ -0,0 +1,96 @@
+
+// ResolveOidcIdentityResponse is the response from the ResolveOidcIdentity RPC.
+message ResolveOidcIdentityResponse {
+    // The principal resolved from the OIDC token.
```
Regenerated

nguyen-andrew commented on proto/redpanda/core/admin/v2/security.proto:
```diff
@@ -0,0 +1,96 @@
+    // The principal resolved from the OIDC token.
+    string principal = 1;
+
+    // The Unix timestamp of the token's expiry.
```
Regenerated

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/2577d2f52fddc33954689caa21febe90bacba957..b7104cf541db7694520fe729166c6385c1dfb71e) - needed to re-generate ducktape protos with updated proto documentation.

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -11,13 +11,70 @@
+        vlog(
+          securitylog.warn,
+          "Failed to authenticate OIDC token: {}",
+          res.has_error() ? res.error().message() : "no value");
```
nit: "unknown" rather than "no value"

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -11,13 +11,70 @@
+          "Failed to authenticate OIDC token: {}",
+          res.has_error() ? res.error().message() : "no value");
+
+        throw serde::pb::rpc::internal_exception(
```
This can be an unauthenticated_exception

michael-redpanda commented on tests/rptest/tests/redpanda_oauth_test.py:
```diff
@@ -659,7 +659,131 @@ def has_group():
+            }
+        )
+
+        self.rpk.create_topic(EXAMPLE_TOPIC)
+        expected_topics = set([EXAMPLE_TOPIC])
+        wait_until(
+            lambda: set(self.rpk.list_topics()) == expected_topics, timeout_sec=10
+        )
```
chore: always supply an `err_msg` to `wait_until`

michael-redpanda reviewed (commented):
looks really nice!

rockwotj commented on proto/redpanda/core/admin/v2/security.proto:
```diff
@@ -0,0 +1,96 @@
+    // The principal resolved from the OIDC token.
+    string principal = 1;
+
+    // The Unix timestamp of the token's expiry.
```
Calling this a Unix timestamp is a bit misleading. It is defined as seconds+nanos since unix epoch, but I think it's simpler to remove the "Unix" word

rockwotj reviewed (commented):
Love how this has been simplified since the first pass. Just one additional nit on top of Mike's feedback.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/b7104cf541db7694520fe729166c6385c1dfb71e..313054841db25e0e86715e17271021e1b309f2a7):
- Addressing PR comments
- Rebase on dev to update to latest changes to `application.cc`.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/1024d2106b53537966edf6fc46ec981a7e89bb99..f8d54f5dc0e0b02cac31c1967644b598793ace30): accidentally forgot to include changes to `tests/rptest/clients/admin/proto/redpanda/core/common/v1/tls_pb2.py` that the `regenerate_ducktape_protos.sh` script generated.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/f8d54f5dc0e0b02cac31c1967644b598793ace30..17b7d07873da341d5534dd260861b86f410528bd) to use `datetime.timezone.utc` instead of `datetime.UTC` (python type checker seems run a version that doesn't have `datetime.UTC`).

rockwotj reviewed (commented):
LGTM, will leave the final approval to Mike

michael-redpanda reviewed (approved):
lgtm!
